src/objective/rank_objective.hpp

```diff
@@ -199,15 +222,26 @@ class LambdarankNDCG : public RankingObjective {
         // get delta NDCG
         double delta_pair_NDCG = dcg_gap * paired_discount * inverse_max_dcg;
         // regular the delta_pair_NDCG by score distance
-        if (norm_ && best_score != worst_score) {
+        if ((norm_ || unbiased_) && best_score != worst_score) {
           delta_pair_NDCG /= (0.01f + fabs(delta_score));
         }
         // calculate lambda for this pair
         double p_lambda = GetSigmoid(delta_score);
         double p_hessian = p_lambda * (1.0f - p_lambda);
+
+        int debias_high_rank = static_cast<int>(std::min(high, truncation_level_ - 1));
+        int debias_low_rank = static_cast<int>(std::min(low, truncation_level_ - 1));
```

Comment:
Hi I am trying to investigate what the issue is with the biases for the truncation cutoff spot that I'm observing and I'm wondering if the thresholding here is the culprit.

Within the context of this implementation, the case where `debias_high_rank` gets the value `truncation_level - 1` is when `high > truncation_level - 1`, i.e. `sorted_idx[high_rank] > truncation_level - 1` i.e. `sorted_idx[j] > truncation_level - 1`. 

Correct me if I'm wrong but the only case the above happens is when `j` is the `high_rank`, and `j` starts from `i+1` which can take a `>truncation_level` value right?

I was wondering if it's better to check the truncation level for both high and low ranks and only update the position biases when both are lower than `truncation_level`, instead of thresholding their values.

As I'm looking for the update to introduce the position column as a metadata value, what I'm thinking is along the lines of:

```c++
const data_size_t high = sorted_idx[high_rank];
const data_size_t low = sorted_idx[low_rank];
data_size_t high_position, low_position;
if (unbiased_) {
  // record_positions_ is a vector of length data_size, with the position of each 
  // record in the data. start is the offset for the current query
  high_position = record_positions_[start + sorted_idx[high_rank]];
  low_position = record_positions_[start + sorted_idx[low_rank]];
} else {
  high_position = high;
  low_position = low;
}

if (unbiased_) {
  double p_cost = log(1.0f / (1.0f - p_lambda)) * delta_pair_NDCG;

  // more relevant (clicked) gets debiased by less relevant (unclicked), only if within truncation levels
  if (high_position <= truncation_level_ && low_position <= truncation_level_) {
    i_costs_buffer_[tid][high_position] += p_cost / j_biases_pow_[low_position];
    j_costs_buffer_[tid][low_position] += p_cost / i_biases_pow_[high_position];  // and vice versa
  }
}
// By default we set values of 1.0 as no-ops
double i_bias_pow = 1.0;
double j_bias_pow = 1.0;
// We only use actual bias values if they are both within the truncation limits
if (high_position <= truncation_level_ && low_position <= truncation_level_) {
  i_bias_pow = i_biases_pow_[high_position];
  j_bias_pow = j_biases_pow_[low_position];
}
// update, either with 1.0 values if at least one of data points ended up outside the truncation threshold or the actual biases
p_lambda *= -sigmoid_ * delta_pair_NDCG / i_bias_pow / j_bias_pow;
p_hessian *= sigmoid_ * sigmoid_ * delta_pair_NDCG / j_bias_pow / j_bias_pow;
```

Does this make sense? The main suggestion is to not "clamp" the bias positions between `[0, truncation_level - 1]` but rather not perform bias updates when one of the document-pair items falls outside the truncation level (making all bias positions valid).

It's a bit unclear to me still the difference between `high/low` and `high_rank/low_rank`. IIRC We sort records in a query by their score and to establish high/low rank we compare their label values.

What I want to ensure is that: since my `record_positions_` will have the original data order, the extracted position corresponds to the expected record, which I think is what we accomplish by `record_positions_[start + sorted_idx[high_rank]]`.